fmt/src/formatter.rs

```diff
@@ -0,0 +1,561 @@
+                write!(self, "{}", item)?;
+
+                if i != items.len() - 1 {
+                    writeln!(self, "{}", separator.trim_end())?;
+                }
+            }
+        } else {
+            write!(self, "{}", items.join(separator))?;
+        }
+
+        Ok(())
+    }
+
+    fn visit_to_string(
+        &mut self,
+        visitable: &mut impl Visitable,
+    ) -> Result<String, Box<dyn std::error::Error>> {
+        self.bufs.push(FormatBuffer::default());
+        Visitable::visit(visitable, self)?;
+        let buf = self.bufs.pop().unwrap();
+
+        Ok(buf.w)
+    }
+
+    /// Returns number of blank lines between two parts of source code
+    fn blank_lines<T: LineOfCode>(&self, a: &mut T, b: &&mut T) -> usize {
+        return self.source[a.loc().2..b.loc().1].matches('\n').count()
+    }
+}
+
+impl<'a, W: Write> Write for Formatter<'a, W> {
+    fn write_str(&mut self, s: &str) -> std::fmt::Result {
+        let (level, current_line, pending_indent, w): (_, _, _, &mut dyn Write) =
+            if let Some(buf) = self.bufs.last_mut() {
+                (buf.level, &mut buf.current_line, &mut buf.pending_indent, &mut buf.w)
+            } else {
+                (self.level, &mut self.current_line, &mut self.pending_indent, self.w)
+            };
+
+        if *pending_indent {
+            let indent = " ".repeat(self.config.tab_width * level);
+            IndentWriter::new(&indent, w).write_str(s)?;
+        } else {
+            w.write_str(s)?;
+        }
+        *current_line += s.len();
+
+        *pending_indent = s.ends_with('\n');
+        if *pending_indent {
+            *current_line = 0;
+        }
+
+        Ok(())
+    }
+}
+
+// Traverse the Solidity Parse Tree and write to the code formatter
+impl<'a, W: Write> Visitor for Formatter<'a, W> {
+    fn visit_source(&mut self, loc: Loc) -> VResult {
+        let source = String::from_utf8(self.source.as_bytes()[loc.1..loc.2].to_vec())?;
+        let mut lines = source.splitn(2, '\n');
+
+        write!(self, "{}", lines.next().unwrap())?;
+        if let Some(remainder) = lines.next() {
+            // Call with `self.write_str` and not `write!`, so we can have `\n` at the beginning
+            // without triggering an indentation
+            self.write_str(&format!("\n{}", remainder))?;
+        }
+
+        Ok(())
+    }
+
+    fn visit_stray_semicolon(&mut self) -> VResult {
+        write!(self, ";")?;
+
+        Ok(())
+    }
+
+    fn visit_newline(&mut self) -> VResult {
+        writeln!(self)?;
+
+        Ok(())
+    }
+
+    fn visit_source_unit(&mut self, source_unit: &mut SourceUnit) -> VResult {
+        // TODO: do we need to put pragma and import directives at the top of the file?
+        // source_unit.0.sort_by_key(|item| match item {
+        //     SourceUnitPart::PragmaDirective(_, _, _) => 0,
+        //     SourceUnitPart::ImportDirective(_, _) => 1,
+        //     _ => usize::MAX,
```

Comment:
Not necessarily, but I'd say it's a good enough practice that I'd always do it, yes